qiskit/pulse/utils.py

```diff
@@ -76,13 +82,10 @@ def format_parameter_value(
                 "converted in a similar fashion to avoid the use of complex parameters.",
                 PendingDeprecationWarning,
             )
-
-        return evaluated
-    except TypeError:
-        # value is not assigned
-        pass
-
-    return operand
+            return evaluated
+    if evaluated.is_integer():
+        return int(evaluated)
```

Comment:
Do you think this is still important? In the old code, the number was always converted from int to float so this was needed to convert the input ints back into ints. With your changes we now know that the input was not an int so it might be okay to leave it as a float. Maybe some users rely on the casting. `is_integer` seems fragile combined with round though. Something that tests for almost an integer might be better.

Reply:
Yes. If the expression is something like 1.00000000001 (could happen if we numerically compute pulse duration), this object is the float type in early type check, and rounded to an integer and typecasted here. This is covered by the unit tests.

Reply:
Probably very large integer with tiny rounding error might be worth testing.

Reply:
Just to check -- do you think `is_integer` is good enough here or should it be

```suggestion
    if np.isclose(evaluated % 1, 0):
        return int(evaluated)
```

to account for rounding error more leniently? Part of the original problem was that `np.round` did not always return something that would pass `is_integer` for an integer input.

Reply:
That's neat. Updated in [8ed6192](https://github.com/Qiskit/qiskit/pull/11972/commits/8ed619275c9f5efd50702587d1c8073a7b0fc005). Thanks!

Reply:
I did some testing because I was still worried about rounding errors. What I found in practice was that `is_integer` always matched `numpy.isclose(val % 1, 0)`, so numpy round must reliably round up above the nearest integer?

I undid your most recent change since `is_integer` seems clearer if there is no computational benefit to the modulus method.

Here is my script for reference. I found that casting to float, rounding, and casting back consistently matched the original number for numbers below `2**52`, which makes sense since that is the precision of the significand in a 64 bit float. I got some surprising results for the fraction of results that did not match the input but that might be because my intuition was off for random numbers and exponential ranges.

```python
import numpy


bad_counts = 0
min_bad = float("inf")
max_err = 0
max_nums = 100_000
total = 0
decimals = 1
max_int = 2**62
max_num = 0
for idx in range(10_000):
    start = numpy.random.randint(0, max_int, size=(10_000,))
    cast = start.astype(numpy.complex128)
    rounded = numpy.round(cast, decimals=decimals).astype(float)
    quotient, remainder = numpy.divmod(rounded, 1)
    # print(f"Round {idx}")
    int_checks = numpy.isclose(remainder, 0)
    # print(f"All close to ints still: {numpy.all(int_checks)}")
    final = quotient.astype(int)
    changed = start != final
    py_is_int = numpy.array([float(f).is_integer() for f in rounded])
    print(f"Not is_int:   {sum(numpy.logical_not(py_is_int))}")
    print(f"Not is_close: {sum(numpy.logical_not(int_checks))}")
    if not numpy.all(py_is_int == int_checks):
        print("is_integer differs from divmod")
    # print(f"int cast is equal: {numpy.all(changed)}")
    if numpy.all(int_checks) and numpy.any(changed):
        for start_i, final_i in zip(start[changed], final[changed]):

            pulse = int(numpy.real(numpy.round(complex(start_i), decimals=decimals)).item())
            # print("Bad input:")
            # print(f"Input:          {start_i}")
            # print(f"Output (numpy): {final_i}")
            # print(f"Output (pulse): {pulse}")
            if pulse != final_i:
                print(f"Pulse diff: {final_i}, {pulse}")
                break
        bad_counts += len(changed)
        min_bad = min(min_bad, min(start[changed]))
        max_err = max(max_err, max(abs(final[changed] - start[changed])))
    max_num = max(max_num, max(start))

    total += len(start)
    if total >= max_nums:
        break

print(f"Minimum bad input: {min_bad} ({min_bad:0.4g}), log2={numpy.log2(min_bad)}")
print(f"Maximum error: {max_err}")
print(f"Bad counts: {bad_counts} / {total}")
print(f"Maximum number processed: {max_num} ({max_num:0.4g}), log2={numpy.log2(max_num)}")
```